### PR TITLE
Bwa meth

### PR DESCRIPTION
**Pull request by dpryan79**
This is an updated version of my last PR (#935) for bwameth from @brentp. Now, bwameth is in its own package and installed the normal python way (or via conda). The test data is also now much smaller (~340K). TravisCI is now happier with everything, which is good. Let me know what else should be modified with this.

### CODE REVIEW

File: packages/package_python_2_7_toolshed_0_4_6/tool_dependencies.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        <repository name="package_python_2_7_10" owner="iuc" prior_installation_required="True" />
+    </package>
+    <package name="toolshed" version="0.4.6">
+        <install version="1.0">
+            <actions>
+                <action type="setup_python_environment">
+                    <repository name="package_python_2_7_10" owner="iuc">
+                        <package name="python" version="2.7.10" />
+                    </repository>
+                    <package md5sum="b0f678e499d5032b9d810271ad26b411">https://pypi.python.org/packages/08/bd/d6259c8d882f6783fdfe69bfaed628afb1ddd291f8a1ac6176d27c62860c/toolshed-0.4.6.tar.gz#md5=b0f678e499d5032b9d810271ad26b411</package>
```
Comment (mvdbeek) [inline packages/package_python_2_7_toolshed_0_4_6/tool_dependencies.xml]: I think you'll have to skip the #md5=b0f678e499d5032b9d810271ad26b411 here.

Comment (dpryan79) [inline packages/package_python_2_7_toolshed_0_4_6/tool_dependencies.xml]: OK, I removed it.

File: tools/bwameth/bwameth.xml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  </requirements>
+  <stdio>
+    <exit_code range="1:" />
+    <exit_code range=":-1" />
+    <regex match="Error:" />
+    <regex match="Exception:" />
+    <regex match="Exception :" />
+  </stdio>
+  <version_command>bwameth.py --version</version_command>
+  <command>
```
Comment (mvdbeek) [inline tools/bwameth/bwameth.xml]: I think the current standard is to use <command detect_errors="aggressive"> in lieu of the <stdio> tag set.
See https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#exit-code-detection

Comment (dpryan79) [inline tools/bwameth/bwameth.xml]: Cool, that's certainly more compact than all the stdio stuff :)

**Comment by mvdbeek** [comment]
Very nice @dpryan79, I'm mirroring the 2 github downloads to the galaxydepot, and then I think this is good to go.

File: packages/package_python_2_7_toolshed_0_4_6/tool_dependencies.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        <repository name="package_python_2_7_10" owner="iuc" prior_installation_required="True" />
+    </package>
+    <package name="toolshed" version="0.4.6">
+        <install version="1.0">
+            <actions>
+                <action type="setup_python_environment">
+                    <repository name="package_python_2_7_10" owner="iuc">
+                        <package name="python" version="2.7.10" />
+                    </repository>
+                    <package md5sum="b0f678e499d5032b9d810271ad26b411">https://pypi.python.org/packages/08/bd/d6259c8d882f6783fdfe69bfaed628afb1ddd291f8a1ac6176d27c62860c/toolshed-0.4.6.tar.gz</package>
```
Comment (mvdbeek) [inline packages/package_python_2_7_toolshed_0_4_6/tool_dependencies.xml]: Can you replace this with https://depot.galaxyproject.org/software/toolshed/toolshed_0.4.6_src_all.tar.gz ?

File: packages/package_python_2_7_bwameth_0_2_0/tool_dependencies.xml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        <install version="1.0">
+            <actions>
+                <action type="setup_python_environment">
+                    <repository name="package_python_2_7_10" owner="iuc">
+                        <package name="python" version="2.7.10" />
+                    </repository>
+                    <repository name="package_python_2_7_toolshed_0_4_6" owner="iuc">
+                        <package name="toolshed" version="0.4.6" />
+                    </repository>
+                    <package md5sum="7d6e2ed375da0fc3d44de3477aa5c087">https://github.com/brentp/bwa-meth/archive/v0.2.0.tar.gz</package>
```
Comment (mvdbeek) [inline packages/package_python_2_7_bwameth_0_2_0/tool_dependencies.xml]: and this with https://depot.galaxyproject.org/software/bwameth/bwameth_0.2.0_src_all.tar.gz ?

**Comment by mvdbeek** [comment]
Sorry for the piecemeal review. It would also be nice if you could use 4 space indentation in bwameth.xml, but I can also do this post-merge if you prefer.

**Comment by dpryan79** [comment]
Thanks for the help @mvdbeek, I've implemented you suggestions!

**Comment by dpryan79** [comment]
Guess I'm not the only one who can't get planemo to use postgresql on Travis (thereby causing random sqlite locking issues). I'll close and reopen this, hopefully that won't happen again.

File: tools/bwameth/bwameth.xml [hunk 5]
```diff
@@ -0,0 +1,102 @@
+<tool id="bwameth" name="bwameth" version="0.2.0">
+    <description>Fast and accurate aligner of BS-Seq reads.</description>
+    <requirements>
+        <requirement type="package" version="0.4.6">toolshed</requirement>
```
Comment (bgruening) [inline tools/bwameth/bwameth.xml]: toolshed does need to be here

File: tools/bwameth/bwameth.xml [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        ln -s '$referenceSource.reference' index_dir/genome.fa &&
+        bwameth.py index index_dir/genome.fa &&
+        #set index="index_dir/genome.fa"
+    #else
+        #set index=$referenceSource.index.fields.path
+    #end if
+
+    bwameth.py
+        -t "\${GALAXY_SLOTS:-4}"
+        --reference ${index}
```
Comment (bgruening) [inline tools/bwameth/bwameth.xml]: Can you quote index please.

File: tools/bwameth/bwameth.xml [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                        <validator type="no_options" message="No indexes are available for the selected input dataset"/>
+                    </options>
+                </param>
+            </when>
+        </conditional>
+
+        <conditional name="single_or_paired">
+            <param name="single_or_paired_opts" type="select" label="Is this library mate-paired?">
+                <option value="single">Single-end</option>
+                <option value="paired">Paired-end</option>
```
Comment (bgruening) [inline tools/bwameth/bwameth.xml]: Do we want to support paired_collections?

File: tools/bwameth/bwameth_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python
```
Comment (bgruening) [inline tools/bwameth/bwameth_wrapper.py]: @dpryan79 is this script used somewhere? I think I'm missing something :(

File: tools/bwameth/bwameth.xml (hunk 5 above)
Comment (dpryan79) [inline tools/bwameth/bwameth.xml]: Good catch, that's a hold-over from the previous implementation.

File: tools/bwameth/bwameth.xml (hunk 7 above)
Comment (dpryan79) [inline tools/bwameth/bwameth.xml]: Sure, can you show me an implementation of that somewhere?

File: tools/bwameth/bwameth_wrapper.py (hunk 8 above)
Comment (dpryan79) [inline tools/bwameth/bwameth_wrapper.py]: Guess I forgot to `git rm` that.

File: tools/bwameth/bwameth.xml (hunk 7 above)
Comment (bgruening) [inline tools/bwameth/bwameth.xml]: This one? https://github.com/galaxyproject/tools-devteam/blob/master/tools/bowtie2/bowtie2_wrapper.xml#L213

File: tools/bwameth/bwameth_wrapper.py (hunk 8 above)
Comment (bgruening) [inline tools/bwameth/bwameth_wrapper.py]: :)

File: tools/bwameth/bwameth.xml (hunk 7 above)
Comment (dpryan79) [inline tools/bwameth/bwameth.xml]: I just pushed something that should hopefully implement this, please have a look and let me know if I borked it :P

Comment (bgruening) [inline tools/bwameth/bwameth.xml]: You could steal a test case from https://github.com/galaxyproject/tools-devteam/blob/master/tools/tophat2/tophat2_wrapper.xml#L370 :)

Comment (dpryan79) [inline tools/bwameth/bwameth.xml]: Guess that was easy enough then.

Comment (bgruening) [inline tools/bwameth/bwameth.xml]: :)

**Comment by bgruening** [comment]
Should be online in 4 min. Thanks again!

**Comment by bgruening** [comment]
Anyone up to create a data manager for this one?

**Comment by dpryan79** [comment]
Thanks @bgruening and @mvdbeek

**Pull request merged**